src/lib/core/Global.h

```diff
@@ -97,6 +111,31 @@ class Global
     ~Global()          = default;
 
 private:
+#if CHIP_CONFIG_DYNAMIC_GLOBALS
+    T & _get() { return *reinterpret_cast<T *>(create(static_cast<T *>(nullptr))); }
+    template <class U, std::size_t N>
+    static U * create(U (*value)[N])
+    {
+#if CHIP_CONFIG_GLOBALS_NO_DESTRUCT
+        static U * ptr = new U[N];
+        return ptr;
```

Comment:
There is a slight setback with this:

Platform::MemoryAlloc, MemoryFree & cie will VerifyOrDieWithoutLogging if used after chip::Platform::MemoryShutdown() or before chip::Platform::MemoryInit().

IIUC, Darwin is the only platform that has CHIP_CONFIG_GLOBALS_LAZY_INIT enabled, in which case __cxa_atexit((f), (p), &__dso_handle) is used to call the destructor of the Global<T>.

However, chip::Platform::MemoryShutdown() happens inside main, before any of the Global<T>s are destroyed, so... if T or any member of T needs to chip::Platform::MemoryFree() something from it's destructor it will crash.

@bzbarsky-apple, is my understanding correct? Any ideas how to mitigate this? I've now moved a bunch of large classes into Global<T>s and I'm starting to wonder if that was good idea in the first place...

Reply:
oh, right. Using `Platform::New` isn't then a good idea in the case of globals. In such a case, I'm fine with the current code, so please only address the comment about `static` local variables.

Reply:
@Damian-Nordic my concern is it might be already broken. In case any member of a member of the classes that I converted to Global<T>s has a destructor that use Platform::Memory*() that would lead to a crash (at least on Darwin).

Even if I change the destroy functions to not use the platform memory allocator, nothing is preventing destructors from using them.

Well, now that I think of it some more, destructors of global variables would have caused that issue even without my changes, so the conversion might be safe in this case.

Reply:
Right, I guess the idea behind `Platform::Memory*` APIs was to use it to allocate objects that are explicitly released using methods like `Shutdown()`. If this is the case, then `Global<>` is simply not a valid use case scenario for `Platform::Memory` so my suggestion was wrong.

Reply:
Thanks @Damian-Nordic!
I've removed the static local and changed the Linux platform config to use CHIP_CONFIG_DYNAMIC_GLOBALS to test the change. I'm not sure how to test this (now opt-in) option unless a platform is actually using it. If that's not OK, any suggestions are welcome.

Reply:
Not sure whether you need something from me here, but using Platform::Memory for something with "static" lifetime indeed does not work.

Reply:
Thanks for confirming @bzbarsky-apple!